packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+to handle starting and stopping the Appium server in Node.
+
+## Standard driver implementation ideas
+
+These are things you will probably find yourself wanting to do when creating a driver.
+
+### Set up state in a constructor
+
+If you define your own constructor, you'll need to call `super` to make sure all the standard state
+is set up correctly:
+
+```js
+constructor(opts) {
+    super(opts);
+    // now do your own thing
+}
+```
+
+### Define and validate accepted capabilities
+
+You can define your own capabilities and basic validation for them. Users will always be able to
+send in capabilities that you don't define, but if they send in capabilities you have explicitly
+defined, then Appium will validate that they are of the correct type (and will check for the
+presence of required capabilities).
+
+If you want to turn capability validation off entirely, set `this.shouldValidateCaps` to `false` in
+your constructor.
+
+To give Appium your validation constraints, set `this.desiredCapConstraints` to a validation object
+in your constructor. Validation objects can be somewhat complex. Here's an example from the
+UiAutomator2 driver:
+
+```js
+{
+  app: {
+    presence: true,
+    isString: true
+  },
+  automationName: {
+    isString: true
+  },
+  browserName: {
+    isString: true
+  },
+  launchTimeout: {
+    isNumber: true
+  },
+}
+```
+
+### Start a session and read capabilities
+
+Appium's `BaseDriver` already implements the `createSession` command, so you don't have to. However
+it's very common to need to perform your own startup actions (launching an app, running some
+platform code, or doing different things based on capabilities you have defined for your driver).
+So you'll probably end up overriding `createSession`. You can do so by defining the method in your
+driver:
+
+```js
+async createSession(jwpCaps, reqCaps, w3cCaps, otherDriverData) {
+    const [sessionId, caps] = super.createSession(w3cCaps);
+    // do your own stuff here
+    return [sessionId, caps];
+}
+```
+
+For legacy reasons, your function will receive old-style JSON Wire Protocol desired and required
+caps as the first two arguments. Given that the old protocol isn't supported anymore and clients
+have all been updated, you can instead only rely on the `w3cCaps` parameter. (For a discussion
+about what `otherDriverData` is about, see the section below on concurrent drivers).
+
+You'll want to make sure to call `super.createSession` in order to get the session ID as well as
+the processed capabilities (note that capabilities are also set on `this.caps`; modifying `caps`
+locally here would have no effect other than changing what the user sees in the create session
+response).
+
+So that's it! You can fill out the middle section with whatever startup logic your driver requires.
+
+### End a session
+
+If your driver requires any cleanup or shutdown logic, it's best to do it as part of overriding the
```

Comment:
it is probably important to mention one has to be careful about throwing exceptions in this method call as we usually want all session deletion steps to be executed before existing the method. So it must be designed in the way that each api call there is error-proof.